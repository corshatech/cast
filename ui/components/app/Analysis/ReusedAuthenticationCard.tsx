/* Copyright 2023 Corsha.
   Licensed under the Apache License, Version 2.0 (the 'License');
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
   http://www.apache.org/licenses/LICENSE-2.0
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an 'AS IS' BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License. */

import React from 'react';
import {
  DataGrid,
  GridColDef,
} from '@mui/x-data-grid';

import { AnalysisOf, ReusedAuthentication } from '@/lib/findings';

import { AnalysisCard, CsvExportButton } from './core';

const columns: GridColDef[] = [
  { field: 'Secret', headerName: 'Secret', width: 400 },
  { field: 'Src IP', headerName: 'Src IP' },
  { field: 'URI', headerName: 'URI', width: 400 },
  { field: 'Dest IP', headerName: 'Dest IP' },
  { field: 'Dest Port', headerName: 'Dest Port' },
];

export const ReusedAuthenticationCard: React.FC<AnalysisOf<ReusedAuthentication>> = ({
  findings,
  reportedAt,
  ...otherProps
}) => {
  const data = (findings ?? []).flatMap((f) => (
    f.data.inRequests.map((rq) => ({
      id: `${f.data.auth}${JSON.stringify(rq)}`,
      'Secret': f.data.auth,
      'Src IP': rq.srcIp,
      'URI': rq.URI,
      'Dest IP': rq.destIp,
      'Dest Port': rq.destPort,
    }))
  ))
  return <AnalysisCard
    reportedAt={reportedAt}
    exportButton={<CsvExportButton
      stripID
      data={data}
      filename={`${reportedAt}-ReusedAuthentication.csv`}
    />}
    {...otherProps}
    noResults={data.length === 0}
  >
    <div style={{height: '400px', width: '100%'}}>
    <DataGrid
      rows={data}
      columns={columns}
      pageSize={10}
      rowsPerPageOptions={[10]}
    />
    </div>
  </AnalysisCard>
}
