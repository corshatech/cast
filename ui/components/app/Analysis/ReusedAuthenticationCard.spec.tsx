/* Copyright 2023 Corsha.
   Licensed under the Apache License, Version 2.0 (the 'License');
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
   http://www.apache.org/licenses/LICENSE-2.0
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an 'AS IS' BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License. */

import { render } from '@testing-library/react';

import { ReusedAuthenticationCard } from './ReusedAuthenticationCard';

const reportedAt = new Date().toISOString();

describe('ReusedAuthenticationCard', () => {
  it('renders all passed props', () => {
    const { asFragment } = render(<ReusedAuthenticationCard
      id="reused-auth"
      title="Test title"
      description="Test description"
      reportedAt={reportedAt}
      severity="high"
      findings={[{
        type: 'reused-auth',
        name: 'Reused Authentication',
        detectedAt: reportedAt,
        severity: 'high',
        data: {
          auth: 'test-data-auth',
          inRequests: [{
            at: reportedAt,
            destIp: '1.1.1.1',
            destPort: '20',
            proto: 'tcp',
            srcIp: '2.2.2.2',
            srcPort: '5129',
            URI: 'https://example.com',
            count: 1,
          }, {
            at: reportedAt,
            destIp: '1.1.1.1',
            destPort: '20',
            proto: 'tcp',
            srcIp: '1.1.1.1',
            srcPort: '5129',
            URI: 'https://example.com',
            count: 1,
          }],
        },
      }]}
    />);
    expect(asFragment()).toMatchSnapshot();
  });
});
