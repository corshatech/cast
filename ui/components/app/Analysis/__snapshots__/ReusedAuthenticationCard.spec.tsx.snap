// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReusedAuthenticationCard renders all passed props 1`] = `
<DocumentFragment>
  <a
    id="reused-auth"
  >
    <div
      aria-labelledby="title"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root shadow-md css-1qo3px0-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root p-0 last:pb-0 css-46bh2p-MuiCardContent-root"
      >
        <div
          class="flex flex-wrap items-start px-2 py-4 md:py-6 md:px-8 justify-between"
        >
          <div
            class="items-center flex w-full md:mb-0"
          >
            <h2
              class="leading-6 text-h2 font-poppins font-bold grow mr-2"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium mb-[3px] pb-[1px] css-1wxstni-MuiSvgIcon-root"
                data-testid="ErrorIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                />
              </svg>
               Test title
            </h2>
            <div
              class="shrink-0"
            >
              <button
                aria-label="download data as CSV"
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                data-mui-internal-clone-element="true"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="FileDownloadOutlinedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M18 15v3H6v-3H4v3c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2v-3h-2zm-1-4-1.41-1.41L13 12.17V4h-2v8.17L8.41 9.59 7 11l5 5 5-5z"
                    />
                  </svg>
                </span>
                Download
              </button>
            </div>
          </div>
          <div
            class="pl-1 mt-2 flex flex-col md:mt-0 md:flex-row md:flex-wrap md:space-x-6"
          >
            <div
              class="flex items-center text-sm text-gray-300"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-calendar mr-1.5 h-3 w-3 flex-shrink-0 text-gray-200"
                data-icon="calendar"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
                  fill="currentColor"
                />
              </svg>
              2023-01-17T13:12:00.000Z
            </div>
            <div
              class="flex items-center text-sm text-gray-300"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium mr-1 h-3 w-3 flex-shrink-0 text-gray-200 css-1wxstni-MuiSvgIcon-root"
                data-testid="ErrorIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                />
              </svg>
              high
            </div>
          </div>
        </div>
        <p
          class="leading-6 text-body1 font-nunito max-w-prose pl-2.5 pr-6 md:pl-9 pb-6"
        >
          Test description
        </p>
        <div
          class="MuiTableContainer-root css-ahxqwc-MuiTableContainer-root"
        >
          <table
            aria-label="collapsible table"
            class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                  scope="col"
                >
                  Secret
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                  scope="col"
                >
                  URI
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                  scope="col"
                >
                  Count
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                  scope="col"
                />
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root overflow-scroll css-apqrd9-MuiTableBody-root"
            >
              <tr
                class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                >
                  test-data-auth
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                >
                  https://example.com
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                >
                  1
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                >
                  <button
                    aria-label="expand row"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="ExpandMoreIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </td>
              </tr>
              <tr
                class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-d9l1z9-MuiTableCell-root"
                  colspan="9"
                />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
